File: src/OfflineLlm.Core/Chat/ChatEngine.cs
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using OfflineLlm.Core.Engine;

namespace OfflineLlm.Core.Chat;

/// <summary>
/// Streams chat completions from a running llama-server instance's OpenAI-compatible
/// /v1/chat/completions endpoint (localhost only).
/// </summary>
public sealed class ChatEngine
{
    private readonly HttpClient _http;

    public ChatEngine(LlamaServerProcess server)
    {
        _http = new HttpClient { BaseAddress = server.BaseAddress };
    }

    public async IAsyncEnumerable<string> StreamReplyAsync(
        IReadOnlyList<ChatMessage> history,
        [System.Runtime.CompilerServices.EnumeratorCancellation] CancellationToken ct = default)
    {
        var payload = new
        {
            model = "local",
            stream = true,
            messages = history.Select(m => new
            {
                role = m.Role switch
                {
                    ChatRole.User => "user",
                    ChatRole.Assistant => "assistant",
                    ChatRole.System => "system",
                    _ => "user",
                },
                content = m.Content,
            }),
        };

        using var request = new HttpRequestMessage(HttpMethod.Post, "v1/chat/completions")
        {
            Content = JsonContent.Create(payload),
        };

        using var response = await _http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, ct);
        response.EnsureSuccessStatusCode();

        await using var stream = await response.Content.ReadAsStreamAsync(ct);
        using var reader = new StreamReader(stream, Encoding.UTF8);

        while (!reader.EndOfStream)
        {
            var line = await reader.ReadLineAsync(ct);
            if (string.IsNullOrWhiteSpace(line) || !line.StartsWith("data:"))
            {
                continue;
            }

            var data = line["data:".Length..].Trim();
            if (data == "[DONE]")
            {
                yield break;
            }

            using var doc = JsonDocument.Parse(data);
            var delta = doc.RootElement
                .GetProperty("choices")[0]
                .GetProperty("delta");

            if (delta.TryGetProperty("content", out var contentEl) && contentEl.ValueKind == JsonValueKind.String)
            {
                var chunk = contentEl.GetString();
                if (!string.IsNullOrEmpty(chunk))
                {
                    yield return chunk;
                }
            }
        }
    }
}

File: app/core/chat_engine.py
"""Streams chat completions from a running llama-server instance's
OpenAI-compatible /v1/chat/completions endpoint (localhost only).
"""

from __future__ import annotations

import json
from typing import Iterator

import requests

from .chat_models import ChatMessage
from .llama_server import LlamaServerProcess

_ROLE_TO_API = {
    "user": "user",
    "assistant": "assistant",
    "system": "system",
}


def stream_reply(server: LlamaServerProcess, history: list[ChatMessage]) -> Iterator[str]:
    payload = {
        "model": "local",
        "stream": True,
        "messages": [
            {"role": _ROLE_TO_API.get(m.role.value, "user"), "content": m.content}
            for m in history
        ],
    }

    with requests.post(server.base_url + "v1/chat/completions", json=payload, stream=True, timeout=120) as response:
        response.raise_for_status()

        for raw_line in response.iter_lines(decode_unicode=True):
            if not raw_line or not raw_line.startswith("data:"):
                continue

            data = raw_line[len("data:"):].strip()
            if data == "[DONE]":
                return

            chunk = json.loads(data)
            delta = chunk["choices"][0]["delta"]
            content = delta.get("content")
            if content:
                yield content
